// vue-material-components/src/components/btn/index.ts
import Component from 'vue-class-component';
import mdIcon from '../icon';

import waveEffect from '../../directives/wave-effect';
import bindBoolean from '../../directives/bind-boolean';

@Component({
    props: {
        type: {
            type: String,
            required: false,
            "default": null,
            twoWay: false
        },
        icon: {
            type: String,
            required: false,
            "default": null,
            twoWay: false
        },
        iconAlignRight: {
            type: Boolean,
            required: false,
            "default": false,
            twoWay: false
        },
        large: {
            type: Boolean,
            required: false,
            "default": false,
            twoWay: false
        },
        disabled: {
            type: Boolean,
            required: false,
            "default": false,
            twoWay: false
        },
        floating: {
            type: Boolean,
            required: false,
            "default": false,
            twoWay: false
        }
    },
    components: {
        mdIcon
    },
    directives: {
        waveEffect,
        bindBoolean
    },
    template: require('./btn.html')
})
export default class Btn {
    private type: string;
    private large: boolean;
    private disabled: boolean;
    private floating: boolean;

    get computedClasses() {
        var classes = {
            'btn-large': this.large,
            'disabled': this.disabled,
            'btn-floating': this.floating
        };
        classes[this.type ? "btn-" + this.type : "btn"] = true;
        return classes;
    }
}
